draft-ietf-dnsop-svcb-httpssvc.md

```diff
@@ -993,41 +1000,41 @@ of {{HSTS}}.
 
 The SVCB "esniconfig" parameter is defined for
 conveying the ESNI configuration of an alternative service.
-The value of the parameter is an ESNIConfig structure {{!ESNI}}
-or the empty string.  ESNI-aware clients SHOULD prefer SVCB/HTTPSSVC RRs with
-non-empty esniconfig.
-
-The parameter value is the ESNIConfig structure {{!ESNI}}
-encoded in {{!base64=RFC4648}} or the empty string.
+The value of the parameter is an ESNIConfig structure {{!ESNI}}.
+In presentation format, the structure is encoded in {{!base64=RFC4648}}.
 The SVCB SvcParamValue wire format is the octet string
 containing the binary ESNIConfig structure.
 
 This parameter MUST NOT appear more than once in a single SvcFieldValue.
 
-
-### Handling a mixture of alternatives not supporting ESNI
+## Client behavior {#esni-client-behavior}
 
 The general client behavior specified in {{client-behavior}} permits clients
 to retry connection with a less preferred alternative if the preferred option
 fails, including falling back to a direct connection if all SVCB options fail.
 This behavior is
 not suitable for ESNI, because fallback would negate the privacy benefits of
-ESNI.
+ESNI.  Accordingly, ESNI-capable clients SHALL implement the following
+behavior for connection establishment.
 
-Accordingly, any connection attempt that uses ESNI MUST fall back only to
-another alt-value that also has the esniconfig parameter.  If the parameter's
-value is the empty string, the client SHOULD connect as it would in the
-absence of any ESNIConfig information.
+1. Perform connection establishment using HTTPSSVC as described in
+   {{client-behavior}}, but do not fall back to the origin's A/AAAA records.
+   If all the HTTPSSVC RRs have esniconfig, and they all fail, terminate
+   connection establishment.
+2. If the client implements Alt-Svc, try to connect using any entries from
+   the Alt-Svc cache.
+3. Fall back to the origin's A/AAAA records if necessary.
 
-For example, suppose a server operator has two alternatives.  Alternative A
-is reliably accessible but does not support ESNI.  Alternative B supports
-ESNI but is not reliably accessible.  The server operator could include a
-full esniconfig value in Alternative B, and mark Alternative A with esniconfig=""
-to indicate that fallback from B to A is allowed.
+As a latency optimization, clients MAY prefetch DNS records for later steps
+before they are needed.
 
-Other clients and services implementing SVCB or HTTPSSVC with esniconfig
-are encouraged to take a similar approach.
+## Deployment considerations
 
+An HTTPSSVC RRSet containing some RRs with esniconfig and some without is
+vulnerable to a downgrade attack.  This configuration is NOT RECOMMENDED.
+Zone owners who do use such a mixed configuration SHOULD mark the RRs with
+esniconfig as more preferred (i.e. smaller SvcFieldPriority) than those
+without.
```

Comment:
Marking it more preferred means ESNI is still not downgrade-protected, right? I.e. the prioirty thing is just to be opportunistically okay? It's probably worth clarifying. This confused me at first.

Reply:
OK, I've added a clarification on this point.